File: advent-of-code/R/2023_01.R
library(tidyverse)


text <- read_lines("data/2023-01.txt")


get_calibration_value <- function(text, regex, replacement = c("x" = "x")) {
  lookahead_pattern <- str_c("(?=(", regex, "))")
  
  text |>
    str_match_all(lookahead_pattern) |>
    map_chr(\(.) str_c(first(.[,2]), last(.[,2]))) |>
    str_replace_all(replacement) |>
    as.integer()
}




answer1 <- get_calibration_value(text, "\\d") |> sum()
print(answer1)




spelling <- c("one", "two", "three", "four", "five", "six", "seven", "eight", "nine")
regex <- str_c("\\d|", str_c(spelling, collapse="|"))
replacement <- seq_along(spelling) |> as.character() |> set_names(spelling)

answer2 <- get_calibration_value(text, regex, replacement) |> sum()
print(answer2)

File: advent-of-code/python/2023_01.py
from aocd import get_data, submit


def first_value(line, m):
  for i, _ in enumerate(line):
    for text, val in m.items():
      if line[i:].startswith(text):
        return val


inp = get_data(day=1, year=2023)
lines = inp.splitlines()
m = {c: int(c) for c in '123456789'}

answer1 = sum(10 * first_value(line, m) + first_value(line[::-1], m) for line in lines)
print(answer1)

submit(answer1, part='a', day=1, year=2023)


# Part 2


m.update({
  'one': 1,
  'two': 2,
  'three': 3,
  'four': 4,
  'five': 5,
  'six': 6,
  'seven': 7,
  'eight': 8,
  'nine': 9,
})
rev_m = {text[::-1]: value for text, value in m.items()}

answer2 = sum(10 * first_value(line, m) + first_value(line[::-1], rev_m) for line in lines)
print(answer2)

submit(answer2, part='b', day=1, year=2023)
